File: lachancuoicgoi_flutter/lib/analysis/common/fuzzy_matcher.dart
class FuzzyMatcher {
  FuzzyMatcher._();

  static int levenshtein(
    String a,
    String b, {
    int maxDistance = 2,
  }) {
    if ((a.length - b.length).abs() > maxDistance) {
      return maxDistance + 1;
    }

    var previous = List<int>.generate(b.length + 1, (index) => index);
    var current = List<int>.filled(b.length + 1, 0);

    for (var i = 1; i <= a.length; i++) {
      current[0] = i;
      var rowMin = current[0];
      for (var j = 1; j <= b.length; j++) {
        final cost = a.codeUnitAt(i - 1) == b.codeUnitAt(j - 1) ? 0 : 1;
        final value = _min3(
          previous[j] + 1,
          current[j - 1] + 1,
          previous[j - 1] + cost,
        );
        current[j] = value;
        if (value < rowMin) rowMin = value;
      }
      if (rowMin > maxDistance) return maxDistance + 1;
      final tmp = previous;
      previous = current;
      current = tmp;
    }

    return previous[b.length];
  }

  static int damerauLevenshtein(
    String a,
    String b, {
    int maxDistance = 2,
  }) {
    if (a == b) return 0;
    if (a.isEmpty) return b.length.clamp(0, maxDistance + 1).toInt();
    if (b.isEmpty) return a.length.clamp(0, maxDistance + 1).toInt();
    if ((a.length - b.length).abs() > maxDistance) {
      return maxDistance + 1;
    }

    var r0 = List<int>.filled(b.length + 1, 0);
    var r1 = List<int>.generate(b.length + 1, (index) => index);
    var r2 = List<int>.filled(b.length + 1, 0);

    for (var i = 1; i <= a.length; i++) {
      r2[0] = i;
      for (var j = 1; j <= b.length; j++) {
        final cost = a.codeUnitAt(i - 1) == b.codeUnitAt(j - 1) ? 0 : 1;
        var value = _min3(
          r1[j] + 1,
          r2[j - 1] + 1,
          r1[j - 1] + cost,
        );

        if (i > 1 &&
            j > 1 &&
            a.codeUnitAt(i - 1) == b.codeUnitAt(j - 2) &&
            a.codeUnitAt(i - 2) == b.codeUnitAt(j - 1)) {
          final transposition = r0[j - 2] + cost;
          value = value < transposition ? value : transposition;
        }

        r2[j] = value;
      }
      final temp = r0;
      r0 = r1;
      r1 = r2;
      r2 = temp;
    }

    return r1[b.length];
  }

  static String? findClosest(
    String token,
    Iterable<String> candidates, {
    int maxDistance = 2,
  }) {
    String? best;
    var bestDistance = maxDistance + 1;

    for (final candidate in candidates) {
      if ((token.length - candidate.length).abs() > maxDistance) continue;
      final distance = damerauLevenshtein(
        token,
        candidate,
        maxDistance: maxDistance,
      );
      if (distance < bestDistance) {
        best = candidate;
        bestDistance = distance;
        if (distance == 0) return candidate;
      }
    }

    return bestDistance <= maxDistance ? best : null;
  }

  static int _min3(int a, int b, int c) {
    final minAB = a < b ? a : b;
    return minAB < c ? minAB : c;
  }
}

File: lachancuocgoi_kotline/app/src/main/java/com/example/lachancuocgoi/Analysis/common/FuzzyMatcher.kt
package com.example.lachancuocgoi.Analysis.common

/**
 * Fuzzy matching sử dụng Levenshtein distance cho STT errors.
 * Dùng khi Aho-Corasick exact match thất bại → thử matching gần đúng.
 */
object FuzzyMatcher {

    /**
     * Tính khoảng cách Levenshtein (edit distance) giữa 2 chuỗi.
     * Early exit nếu vượt maxDistance.
     */
    fun levenshtein(a: String, b: String, maxDistance: Int = 2): Int {
        if (a == b) return 0
        if (a.isEmpty()) return b.length.coerceAtMost(maxDistance + 1)
        if (b.isEmpty()) return a.length.coerceAtMost(maxDistance + 1)

        val lenDiff = kotlin.math.abs(a.length - b.length)
        if (lenDiff > maxDistance) return maxDistance + 1

        // DP với 2 hàng để tiết kiệm memory
        var prev = IntArray(a.length + 1) { it }
        var curr = IntArray(a.length + 1)

        for (j in 1..b.length) {
            curr[0] = j
            var minVal = j
            for (i in 1..a.length) {
                val cost = if (a[i - 1] == b[j - 1]) 0 else 1
                curr[i] = minOf(
                    curr[i - 1] + 1,       // insert
                    prev[i] + 1,            // delete
                    prev[i - 1] + cost      // replace
                )
                minVal = minOf(minVal, curr[i])
            }
            // Early exit: nếu minVal trong hàng đã vượt maxDistance
            if (minVal > maxDistance) return maxDistance + 1
            val tmp = prev
            prev = curr
            curr = tmp
        }
        return prev[a.length]
    }

    /**
     * Tìm token gần nhất trong danh sách từ khóa.
     * @param token Token cần tìm (đã normalize)
     * @param candidates Danh sách từ khóa ứng viên (đã normalize)
     * @param maxDistance Khoảng cách tối đa cho phép (mặc định 2 cho STT errors)
     * @return Từ khóa khớp gần nhất, hoặc null nếu không tìm thấy
     */
    fun findClosest(token: String, candidates: Collection<String>, maxDistance: Int = 2): String? {
        var bestMatch: String? = null
        var bestDist = maxDistance + 1

        for (candidate in candidates) {
            // Quick length check trước khi tính DP
            if (kotlin.math.abs(token.length - candidate.length) > maxDistance) continue
            val dist = damerauLevenshtein(token, candidate, maxDistance)
            if (dist < bestDist) {
                bestDist = dist
                bestMatch = candidate
            }
        }
        return if (bestDist <= maxDistance) bestMatch else null
    }

    /**
     * Tính khoảng cách Damerau-Levenshtein (Extended Levenshtein).
     * Bao gồm cả transposition (hoán đổi 2 ký tự liền nhau).
     * Rất hữu ích cho STT errors vì tính nature của speech recognition.
     */
    fun damerauLevenshtein(a: String, b: String, maxDistance: Int = 2): Int {
        if (a == b) return 0
        if (a.isEmpty()) return b.length.coerceAtMost(maxDistance + 1)
        if (b.isEmpty()) return a.length.coerceAtMost(maxDistance + 1)

        val lenDiff = kotlin.math.abs(a.length - b.length)
        if (lenDiff > maxDistance) return maxDistance + 1

        val aLen = a.length
        val bLen = b.length

        // DP với full matrix cho Damerau-Levenshtein
        val dp = Array(aLen + 1) { IntArray(bLen + 1) { 0 } }

        for (i in 0..aLen) dp[i][0] = i
        for (j in 0..bLen) dp[0][j] = j

        for (i in 1..aLen) {
            for (j in 1..bLen) {
                val cost = if (a[i - 1] == b[j - 1]) 0 else 1
                dp[i][j] = minOf(
                    dp[i - 1][j] + 1,        // deletion
                    dp[i][j - 1] + 1,        // insertion
                    dp[i - 1][j - 1] + cost  // substitution
                )
                // Transposition (hoán đổi 2 ký tự liền nhau)
                if (i > 1 && j > 1 && a[i - 1] == b[j - 2] && a[i - 2] == b[j - 1]) {
                    dp[i][j] = minOf(dp[i][j], dp[i - 2][j - 2] + cost)
                }
            }
        }

        return dp[aLen][bLen]
    }
}
